### DISCUSSION TITLE
dsl::try_ still seems to fail on errors

### DISCUSSION THREAD

**Discussion by lava** [foonathan/lexy#40] [Q&A] [answered]
I'm not sure if this is a bug or if I'm just misunderstanding how `try_` is supposed to work, but using this sample program on current master:

```
#include <lexy/dsl.hpp>
#include <lexy/parse.hpp>
#include <lexy/input/string_input.hpp>
#include <lexy_ext/report_error.hpp>

#include <string>

namespace dsl = lexy::dsl;

struct foo {
	static constexpr auto rule = dsl::try_(dsl::lit_c<'t'>, dsl::return_);
};

int main() {
	std::string source_line = "x";
	auto input = lexy::string_input<lexy::utf8_encoding>(source_line);
	lexy::validate<foo>(input, lexy_ext::report_error);
}
```

I'd expect `foo` to accept any input. However, instead I get an error because the first rule in the `try_` block fails:

```
$ ./test
error: while parsing foo
     | 
 1: 1| x
     | ^ expected 't'
------
```

**Comment by foonathan** [comment, marked as answer]
Yes, this is the expected behavior.

`dsl::try_()` is used to do error recovery: if an error happens, it is still reported, but parsing continues after doing the optional recovery part: https://lexy.foonathan.net/playground/?id=7zf1WG4oG

If you want to make something optional, use either `dsl::if_(branch)`, which parses the branch if the condition matches, and does nothing otherwise, or `dsl::opt_(branch)`, which produces a special `lexy::nullopt` value if the branch was not taken: https://lexy.foonathan.net/playground/?id=G1ndY9jrb

(Maybe the name `dsl::try_()` isn't ideal)
